### PR TITLE
Producer send offsets tx

### PR DESCRIPTION
**Pull request by germanosin**

### CODE REVIEW

**Review by bluurn — APPROVED** [review]
LGTM

File: spec/broker_spec.rb [hunk 1]
```diff
@@ -134,4 +134,27 @@ def close
       broker.disconnect
     end
   end
+
+  describe "#add_offsets_to_txn" do
+    it "should send offsets to the transaction" do
```
Comment (stasCF) [inline spec/broker_spec.rb]:
```suggestion
    it 'sends offsets to the transaction' do
```

File: spec/broker_spec.rb [hunk 2]
```diff
@@ -134,4 +134,27 @@ def close
...
+      allow(connection).to receive(:send_request)
```
Comment (stasCF) [inline spec/broker_spec.rb:140]:
```suggestion
      expect(connection).to receive(:send_request)
```

File: spec/broker_spec.rb [hunk 3]
```diff
@@ -134,4 +134,27 @@ def close
...
+        transactional_id: 1, producer_id: 2, producer_epoch: 3, group_id: 4
+      )
+      expect(connection).to have_received(:send_request)
+
+      broker.disconnect
+    end
+  end
+
+  describe "#txn_offset_commit" do
+    it "commit transaction" do
```
Comment (stasCF) [inline spec/broker_spec.rb]:
```suggestion
    it 'commits transaction' do
```

File: spec/broker_spec.rb [hunk 4]
```diff
@@ -134,4 +134,27 @@ def close
...
+      allow(connection).to receive(:send_request)
```
Comment (stasCF) [inline spec/broker_spec.rb:152]:
```suggestion
      expect(connection).to receive(:send_request)
```

File: spec/producer_spec.rb [hunk 5]
```diff
@@ -323,6 +324,13 @@
     end
   end
 
+  describe "#send_offsets_to_transaction" do
+    it "should send offsets to transaction manager" do
```
Comment (stasCF) [inline spec/producer_spec.rb]:
```suggestion
    it 'sends offsets to transaction manager' do
```

File: spec/transaction_manager_spec.rb [hunk 6]
```diff
@@ -567,6 +567,38 @@
       end
     end
   end
+
+  describe '#send_offsets_to_txn' do
+    let!(:manager) do
+      described_class.new(
+        logger: logger, cluster: cluster,
```
Comment (stasCF) [inline spec/transaction_manager_spec.rb]:
```suggestion
        logger: logger,
        cluster: cluster,
```

File: spec/transaction_manager_spec.rb [hunk 7]
```diff
@@ -567,6 +567,38 @@
...
+        transactional: true, transactional_id: 'IDID', transactional_timeout: 600
```
Comment (stasCF) [inline spec/transaction_manager_spec.rb]:
```suggestion
        transactional: true,
        transactional_id: 'IDID',
        transactional_timeout: 600
```

File: spec/transaction_manager_spec.rb [hunk 8]
```diff
@@ -567,6 +567,38 @@
...
+          )
+        )
+        allow(transaction_coordinator).to receive(:txn_offset_commit).and_return(
+          Kafka::Protocol::TxnOffsetCommitResponse.new(
+            error_code: 0
+          )
+        )
+      end
+
+      it 'should notify transaction coordinator' do
```
Comment (stasCF) [inline spec/transaction_manager_spec.rb]:
```suggestion
      it 'notifies transaction coordinator' do
```

File: spec/transaction_manager_spec.rb [hunk 9]
```diff
@@ -567,6 +567,38 @@
...
+        logger: logger, cluster: cluster,
+        transactional: true, transactional_id: 'IDID', transactional_timeout: 600
+      )
+    end
+
+    context 'currently in transaction' do
+      before do
+        manager.init_transactions
+        manager.begin_transaction
+        allow(transaction_coordinator).to receive(:add_offsets_to_txn).and_return(
```
Comment (stasCF) [inline spec/transaction_manager_spec.rb:586]:
```suggestion
        expect(transaction_coordinator).to receive(:add_offsets_to_txn).and_return(
```

File: spec/transaction_manager_spec.rb [hunk 10]
```diff
@@ -567,6 +567,38 @@
...
+          Kafka::Protocol::AddOffsetsToTxnResponse.new(
+            error_code: 0
+          )
+        )
+        allow(transaction_coordinator).to receive(:txn_offset_commit).and_return(
```
Comment (stasCF) [inline spec/transaction_manager_spec.rb:591]:
```suggestion
        expect(transaction_coordinator).to receive(:txn_offset_commit).and_return(
```

**Review by stasCF — APPROVED** [review]
lil codestyle changes only

File: spec/producer_spec.rb [hunk 11]
```diff
@@ -324,10 +324,54 @@
     end
   end
 
-  describe "#send_offsets_to_transaction" do
+  describe "#send_offsets_to_transaction_batch" do
+    let(:batch) do
+      double(
+        topic: 'some_topic',
+        partition: 'some_partition',
```
Comment (germanosin) [inline spec/producer_spec.rb]: partition is int

File: spec/producer_spec.rb [hunk 12]
```diff
@@ -324,10 +324,54 @@
...
+        last_offset: 1,
```
Comment (germanosin) [inline spec/producer_spec.rb]: let's use random int here 0 - 2^31

File: spec/producer_spec.rb [hunk 13]
```diff
@@ -324,10 +324,54 @@
...
+        leader_epoch: 1
```
Comment (germanosin) [inline spec/producer_spec.rb]: Let's use current timestamp here

File: spec/producer_spec.rb [hunk 14]
```diff
@@ -324,10 +324,54 @@
...
+      expect(transaction_manager).to have_received(:send_offsets_to_txn).with(
+        offsets: {
+          'some_topic' => {
+            'some_partition' => {
+              leader_epoch: 1,
+              offset: 2
+            }
+          }
+        },
+        group_id: 1
```
Comment (germanosin) [inline spec/producer_spec.rb]: Let's use uuid string for group id

File: spec/producer_spec.rb [hunk 15]
```diff
@@ -324,10 +324,54 @@
...
+      )
+    end
+  end
+
+  describe "#send_offsets_to_transaction_message" do
+    let(:message) do
+      double(
+        topic: 'some_topic',
+        partition: 'some_partition',
```
Comment (germanosin) [inline spec/producer_spec.rb]: partition is int

File: lib/kafka/fetched_message.rb [hunk 16]
```diff
@@ -8,10 +8,14 @@ class FetchedMessage
...
     attr_reader :partition
 
-    def initialize(message:, topic:, partition:)
+    # @return [Integer] partition leader epoch
+    attr_reader :leader_epoch
+
+    def initialize(message:, topic:, partition:, leader_epoch: nil)
       @message = message
-      @topic = topic
+      @topic = topic
```
Comment (stasCF) [inline lib/kafka/fetched_message.rb]:
```suggestion
      @topic = topic
```

**Review by stasCF — APPROVED** [review]

**Pull request merged**